tachyon/math/finite_fields/cyclotomic_multiplicative_subgroup.h

```diff
@@ -0,0 +1,129 @@
+// Copyright 2022 arkworks contributors
+// Use of this source code is governed by a MIT/Apache-2.0 style license that
+// can be found in the LICENSE-MIT.arkworks and the LICENCE-APACHE.arkworks
+// file.
+
+#ifndef TACHYON_MATH_FINITE_FIELDS_CYCLOTOMIC_MULTIPLICATIVE_SUBGROUP_H_
+#define TACHYON_MATH_FINITE_FIELDS_CYCLOTOMIC_MULTIPLICATIVE_SUBGROUP_H_
+
+#include <tuple>
+#include <vector>
+
+#include "tachyon/base/containers/adapters.h"
+#include "tachyon/math/finite_fields/finite_field.h"
+
+namespace tachyon::math {
+namespace internal {
+
+SUPPORTS_UNARY_IN_PLACE_OPERATOR(FastCyclotomicSquare);
+SUPPORTS_UNARY_IN_PLACE_OPERATOR(FastCyclotomicInverse);
+
+}  // namespace internal
+
+// clang-format off
+// Fields that have a cyclotomic multiplicative subgroup, and which can
+// leverage efficient inversion and squaring algorithms for elements in this
+// subgroup. If a field has multiplicative order pᵈ - 1, the cyclotomic
+// subgroups refer to subgroups of order φₙ(p), for any n < d, where φₙ is the
+// [n-th cyclotomic polynomial](https://en.wikipedia.org/wiki/Cyclotomic_polynomial).
+// clang-format on
+template <typename F>
```

Comment:
F is a PrimeField type right? These days I'm thinking if it's a good naming since we now also have `FiniteField`.

Reply:
Let's say for `Fp2<Config>`. It inherits as follows:

`Fp2<Config>`  -> `QuadraticExtnesionField<Fp2<Config>>`
`QuadraticExtensionField<Derived>` -> `CyclotomicMultiplicativeSubgroup<Derived>`
`CyclotomicMultiplicativeSubgroup<F>`  -> `FiniteField<F>`

So this is what I meant to say `Fp2<Config> == Derived == F`